metadata/src/main/java/org/apache/kafka/metadata/storage/Formatter.java

```diff
@@ -0,0 +1,505 @@
+                    MetadataVersion.FEATURE_NAME, releaseVersion);
+                Files.createDirectories(Paths.get(writeLogDir));
+                BootstrapDirectory bootstrapDirectory = new BootstrapDirectory(writeLogDir, Optional.empty());
+                bootstrapDirectory.writeBinaryFile(bootstrapMetadata);
+                if (directoryTypes.get(writeLogDir).isDynamicMetadataDirectory()) {
+                    writeDynamicQuorumSnapshot(writeLogDir,
+                        initialVoters.get(),
+                        featureLevels.get(KRaftVersion.FEATURE_NAME),
+                        controllerListenerName);
+                }
+            });
+            copier.setWriteErrorHandler((errorLogDir, e) -> {
+                throw new FormatterException("Error while writing meta.properties file " +
+                        errorLogDir + ": " + e);
+            });
+            copier.writeLogDirChanges();
+        }
+    }
+
+    enum DirectoryType {
+        LOG_DIRECTORY,
+        STATIC_METADATA_DIRECTORY,
+        DYNAMIC_METADATA_NON_VOTER_DIRECTORY,
+        DYNAMIC_METADATA_VOTER_DIRECTORY;
+
+        String description() {
+            switch (this) {
+                case LOG_DIRECTORY:
+                    return "data directory";
+                case STATIC_METADATA_DIRECTORY:
+                    return "metadata directory";
+                case DYNAMIC_METADATA_NON_VOTER_DIRECTORY:
+                    return "dynamic metadata directory";
+                case DYNAMIC_METADATA_VOTER_DIRECTORY:
+                    return "dynamic metadata voter directory";
+            }
+            throw new RuntimeException("invalid enum type " + this);
+        }
+
+        boolean isDynamicMetadataDirectory() {
+            return this == DYNAMIC_METADATA_NON_VOTER_DIRECTORY ||
+                this == DYNAMIC_METADATA_VOTER_DIRECTORY;
+        }
+
+        static DirectoryType calculate(
+            String logDir,
+            String metadataLogDirectory,
+            int nodeId,
+            Optional<DynamicVoters> initialVoters
+        ) {
+            if (!logDir.equals(metadataLogDirectory)) {
+                return LOG_DIRECTORY;
+            } else if (!initialVoters.isPresent()) {
+                return STATIC_METADATA_DIRECTORY;
+            } else if (initialVoters.get().voters().containsKey(nodeId)) {
+                return DYNAMIC_METADATA_VOTER_DIRECTORY;
+            } else {
+                return DYNAMIC_METADATA_NON_VOTER_DIRECTORY;
+            }
+        }
+    }
+
+    static void writeDynamicQuorumSnapshot(
+        String writeLogDir,
+        DynamicVoters initialVoters,
+        short kraftVersion,
+        String controllerListenerName
+    ) {
+        File parentDir = new File(writeLogDir);
+        File clusterMetadataDirectory = new File(parentDir, String.format("%s-%d",
+                CLUSTER_METADATA_TOPIC_PARTITION.topic(),
+                CLUSTER_METADATA_TOPIC_PARTITION.partition()));
```

Comment:
Okay. Unfortunate. This is computed by `UnifiedLog.logDirName` but that is part of the `core` module.